R/req-perform.R

```diff
@@ -34,8 +38,22 @@
 #' @examples
 #' request("https://google.com") %>%
 #'   req_perform()
-req_perform <- function(req, path = NULL, verbosity = getOption("httr2_verbosity", 0L)) {
+req_perform <- function(
+      req,
+      path = NULL,
+      verbosity = getOption("httr2_verbosity", 0L),
+      mock = getOption("httr2_mock", NULL)
+  ) {
   check_request(req)
+
+  if (!is.null(mock)) {
+    mock <- as_function(mock)
+    mock_resp <- mock(req)
+    if (!is.null(mock_resp)) {
+      return(mock_resp)
+    }
+  }
+
```

Comment:
I wonder if the mock should instead be run after the next few lines since they appear to modify `req` (though I haven't looked at what they do). I vaguely remember some having to handle some subtlety in `httptest` because of default values getting set somewhere deep down that weren't a part of the `request` object earlier. It seems that here, for example, you wouldn't be able to test anything about auth or oauth tokens being set in the request because that doesn't get added to the request object until after the mock is run.

Reply:
I don't think `req_verbosity()` is meaningful for mocking; it's only used for debugging live requests.

And the challenge with OAuth is that signing process might need to perform its own requests (e.g. in order to refresh a token) — you'd then have to be careful to also mock those requests, which is going to be hard because they only occur occasionally.